### PR TITLE
Identity management proposal

### PR DESCRIPTION
**Pull request by kyleterry**
[Platform comparison rendered](https://github.com/puppetlabs/relay-rfcs/blob/features/identity-management/content/0012-identity-management/platform-comparison.md)
[RFC rendered](https://github.com/puppetlabs/relay-rfcs/blob/features/identity-management/content/0012-identity-management/rfc.md)

### CODE REVIEW

**Review by vine77 — APPROVED** [review]
+1 for the GUI on either option. Seems like Auth0 has a couple options for theming, though the new [Universal Login](https://auth0.com/docs/universal-login) experience's [Page Templates](https://auth0.com/docs/universal-login/new-experience/universal-login-page-templates) don't provide for CSS customization (just custom logo, primary color, background color, and wrapper HTML). @Sigler might have something to say about that, but it doesn't bother me too much and I don't think it should be a deciding factor.

**Comment by woodburndesigns** [comment]
@kyleterry a big question from me is how do we do this all as iteratively as possible. Could we keep our current RBAC system and migrate to theirs over time. Can we see value after the first sprint? How can we prevent this from being a big bang release in January?

**Comment by rick-a-lane-ii** [comment]
Side note: the `Deliverables` section was added to split up the design into functional iterations. It specifically should avoid representing fine-grained development tasks, which can be captured separately in tickets/issues as necessary or warranted. The intention is to break apart the work into at least two major deliveries that are testable and demoable, and have some functional value. It's also meant to reinforce getting usable functionality into staging/production iteratively for other engineers that may be working on or waiting on parts of the system. These should align more with the Sprint "goals" we've been defining.

File: content/0012-identity-management/rfc.md
Comment (impl) [inline content/0012-identity-management/rfc.md]: Unfortunately we use the term `tenant` in Relay Core to refer to a workflow (and, for what it's worth, `domain` to refer to an account). Is there any other name we could possibly choose here instead?

File: content/0012-identity-management/rfc.md
Comment (impl) [inline content/0012-identity-management/rfc.md]: Is our password hashing (bcrypt) compatible with IdPs we're considering?

File: content/0012-identity-management/rfc.md
Comment (impl) [inline content/0012-identity-management/rfc.md:102]: Do we want to cache any of the other data from upstream? For example I expect we'll at least want something like a user's name/email address so we can render it in tables, notifications, etc. Otherwise we'll have to query the IdP management API every time we want to use user information.

File: content/0012-identity-management/rfc.md
Comment (impl) [inline content/0012-identity-management/rfc.md:126]: I think this is OK, but if we can kick the can down the road on this and just use our existing RBAC system at first, it feels like it would be easier to make sure we get this done incrementally.

File: content/0012-identity-management/rfc.md
Comment (impl) [inline content/0012-identity-management/rfc.md]: I assume we'll need some other fields here, too, like what plan an account is on. In fact it might be nice to understand the process of managing accounts at the IdP level here more generally. For example, I know some things that we potentially want to do are:

- Allow a user to be part of multiple accounts
- Restrict the number of users in an account
- Require users for a certain account to have a certain email address suffix/domain
- Require users for a certain account to use a particular auth flow (SAML/ADFS)

In fact, from reading Auth0's documentation on shared users in a B2B context, it seems like their recommendation is not to set up a tenant per (what we currently call an) account, but rather to use metadata to do the mapping instead. This [architecture document](https://auth0.com/docs/architecture-scenarios/implementation/b2b/b2b-provisioning) is very good and particularly relevant to us, as is pretty much everything else in the B2B section. Worth reading.

**Review by impl — COMMENTED** [review]
A couple of thoughts for you, but in general, I think this analysis is pretty solid and the recommendations make sense. I would like us to settle on an IdP before we merge this, if possible, and also to make sure we understand the client-side implications of switching to SSO/OAuth flow for login/signup.

As Rick and Geoff mentioned, in the engineering-level explanation, I'd at least like to see a very early deliverable that doesn't go into the complexities of user migration, moving account terminology around, or trying to shift RBAC. Just being able to sign up and use Relay with an IdP in parallel with our existing database would be a great first step, even if it does some sketchy stuff like force an account per user with no support for invites.

I'm also a little hesitant to schedule downtime for migrating users, at least if we choose Auth0, given that automatic user migration without password resets is a supported mechanism.

File: content/0012-identity-management/rfc.md
Comment (kyleterry) [reply content/0012-identity-management/rfc.md]: Yep! It's recommended by Auth0 and Keycloak to import bcrypt hashes.

File: content/0012-identity-management/rfc.md
Comment (kyleterry) [reply content/0012-identity-management/rfc.md:102]: The answer is most likely, but we will add these after the first iteration. We can add non-sensitive metadata to the JWT claims for the time being.

**Pull request merged**
